### PR TITLE
Allow JS action after module import

### LINKED ISSUE #31274: Issues when upgrading a module used as external source
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When we try to update a module implementing the hook `actionListModules`, if there is any change in the service defintions, the upgrade fails

For example, ps_mbo is an external source module, it registers `actionListModules`. That hook returns the fresh data about modules from external APIs
During the upgrade process, we are call ```$this->moduleRepository->getModule($name);```
This call will request external sources to get fresh data.
In the case it's the module MBO itself which is updating, any changes on service definitions of the new version will lead to a failure

Another issue during module upgrade happens when the module listens to event `ModuleManagementEvent::UPGRADE`. The listeners and subcribers are stored on Sf Container build as Closure (see RegisterListenerPass) and until the process ends. So whatever is done in the upgrade process (cache clean, ...) the subscriber will remain the same. If we use some module services in that subscriber and they have changed, ClassLoader will yell because service definitons changed.<br>To help module developpers to avoid this case, we add a "JS hook" to help modules to execute tasks after a module import

### Expected behavior

_No response_

### Steps to reproduce

1 - Install mbo 4.0.1
2 - Try to upgrade to 4.3.0 using the button "Install a module"
3 - You'll have an error saying that service `AddonsUser` changed defintion

OR

1 - Install v1.0.0 of the example module using the button "Install a module" [prestashopexamplemodule_1.0.0.zip](https://github.com/PrestaShop/PrestaShop/files/10814931/prestashopexamplemodule_1.0.0.zip)
2 - Install v1.0.1 of the example module using the button "Install a module" [prestashopexamplemodule_1.0.1.zip](https://github.com/PrestaShop/PrestaShop/files/10814943/prestashopexamplemodule_1.0.1.zip) => this will perform an upgrade 

==> You'll have an erreur about array_merge with a value that is not an array

### PrestaShop version(s) where the bug happened

8.0.x

### PHP version(s) where the bug happened

_No response_

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | We have some issues on module upgrade process<br><br>One of them happens when the module listens to event `ModuleManagementEvent::UPGRADE`. The listeners and subcribers are stored on Sf Container build as Closure (see RegisterListenerPass) and until the process ends. So whatever is done in the upgrade process (cache clean, ...) the subscriber will remain the same. If we use some module services in that subscriber and they have changed, ClassLoader will yell because service definitons changed.<br>To help module developpers to avoid this case, we add a "JS hook" to help modules to execute tasks after a module import
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | ~
| Fixed ticket?     | Partially Fixes #31274

### CODE REVIEW

File: src/Core/Module/ModuleRepository.php
Comment (matthieu-rolland) [inline src/Core/Module/ModuleRepository.php]: Seems like`$modulesFromHook` can be used in a foreach without being defined, maybe you can give it a default `[]` value ?

Comment (matthieu-rolland) [reply src/Core/Module/ModuleRepository.php]: ah it's a draft, my bad !

Comment (sowbiba) [reply src/Core/Module/ModuleRepository.php]: Fixed @matthieu-rolland 

It's more a proposition than a draft
I'm ok with reviews 😉 

I made a draft because I maybe have more things to fix

File: src/Core/Module/ModuleRepository.php
Comment (sowbiba) [inline src/Core/Module/ModuleRepository.php]:
```suggestion
        foreach ($modulesFromHook as $moduleFromHook) {
```

**Comment by sowbiba** [comment]
Testing the event dispatched after upgrade

1 - Install [prestashopexamplemodule_1.0.0_upgrade_event.zip](https://github.com/PrestaShop/PrestaShop/files/10831278/prestashopexamplemodule_1.0.0_upgrade_event.zip) . You'll have in the configration table an entry PRESTASHOPEXAMPLEMODULE_DEBUG with null value

2 - Import the upgrade version 
[prestashopexamplemodule_1.0.1_upgrade_event.zip](https://github.com/PrestaShop/PrestaShop/files/10831280/prestashopexamplemodule_1.0.1_upgrade_event.zip) . The post install event is triggered and the value of PRESTASHOPEXAMPLEMODULE_DEBUG will be : Upgraded at Y/m/d H:i:s

**Comment by sowbiba** [comment]
Testing the Ajax call after upgrade

1 - Install 
[prestashopexamplemodule_1.0.0_js_after_upgrade.zip](https://github.com/PrestaShop/PrestaShop/files/10831582/prestashopexamplemodule_1.0.0_js_after_upgrade.zip) . You'll have in the configration table an entry PRESTASHOPEXAMPLEMODULE_DEBUG with null value

2 - Import the upgrade version
[prestashopexamplemodule_1.0.1_js_after_upgrade.zip](https://github.com/PrestaShop/PrestaShop/files/10831583/prestashopexamplemodule_1.0.1_js_after_upgrade.zip) . You'll have an ajax call after the module import and the value of PRESTASHOPEXAMPLEMODULE_DEBUG will be : Upgraded at Y/m/d H:i:s

**Review by kpodemski — DISMISSED** [review]

**Comment by sowbiba** [comment]
@PrestaShop/prestashop-maintainers Don't forget me 😄

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/module/controller.js]: One question I have about this, I'm ok for the principle as a temporary solution. I was just wondering if this unique modification is enough for all the use cases? It seems like it adds the possibility for import only What about the upgrade, install, all the other actions? Should there be a similar JS action hook? (still temporary until we improve the whole workflow).

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: Oh and since this is supposed to be a temporary solution, we should add a comment to explain this here, and in the Hook description as well.

**Review by matks — DISMISSED** [review]
This pragmatic approach is OK for me 👍

**Review by zuk3975 — DISMISSED** [review]

File: install-dev/data/xml/hook.xml
Comment (jolelievre) [inline install-dev/data/xml/hook.xml]:
```suggestion
      <description>This hook allows to add a JS function to execute after a module import. This will help to perform actions after module upgrade in a different process (this hook is temporary and ill be removed in the next major)</description>
```

**Comment by HanaRebaiQA** [comment]
Hello @sowbiba 

Thanks for your PR. What about @jolelievre 'review ? Tt will be taken into consideration? 
Can i delete the Waiting for QA label ?

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (kpodemski) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: I'm not sure why you want to treat it as a temporary solution. I can think of some other scenarios where this hook could be handy. Sometimes devs wanted to display something at the end of the upgrade process (without reloading the page). It's a good starting point for doing so.

Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: Good question @jolelievre 

Actually the only scenario I see is when some definitions change, so after an upgrade.
In the other cases we can use the eventListeners without problem

We have 2 ways to perform an upgrade
- Upload a zip file 
- Upgrade pushed by an External Modules catalogue (aka Addons or DistribApi)

I maybe just treated the first case ...

Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: @kpodemski You're absolutely right. This JS hook allows to do "anything" after a module import

It's just a hook as any other one and it doesn't disturb the Core import process

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: Ok @kpodemski that makes sense, then we can keep it even later Howver the two step process should be handled by the core in the future, this hook will remain for other use cases then

File: install-dev/data/xml/hook.xml
Comment (jolelievre) [reply install-dev/data/xml/hook.xml]: Not needed then

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: So @sowbiba should we block the QA on this PR until you handle the other use case of upgrade by catalog?

**Comment by sowbiba** [comment]
Because of the @jolelievre and @kpodemski feedbacks, I completely change the approach

Now we dispatch an JS event after the module upgrade. So any module can catch it and perform actions
The event is triggered in these cases
- the module is upgraded using the button in Module manager
- the module is upgraded using import
- many modules are upgraded in a bulk action

I think it's way better like this because
- no hook
- events are already used but not for upgrade, so it's completing a lack
- anyone can do anything just catching the event

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (kpodemski) [inline admin-dev/themes/new-theme/js/pages/module/controller.js]:
```suggestion
```

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (kpodemski) [inline admin-dev/themes/new-theme/js/pages/module/controller.js]: I don't think it's a good place for such code. Can't we use the event emitter that we already have in BO?

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: I let it on purpose to ... do something 

I can clean it ;)

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: Agreed I guess you can use the already exiting `window.BOEvent`

Ideally, we should even use the `window.prestashop.instance.eventEmitter` that is more generic and used in other pages, but the purpose is not to refacto this event system for modules, for now.

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/module/controller.js]: Shouldn't we trigger `'Module Installed'` when upgraded is false? To be consistent

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: I can use the eventEmitter component but it means do the same for all the events https://github.com/PrestaShop/PrestaShop/pull/31285/files#diff-e7328df916ab4f26f79b085040bfe2a0e1c5a4b9f1564365371b78df89ff17d2R228

That means more code, more QA :D :D

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/module/controller.js]: Yes the console.log shouldn't stay here, I guess you can leave a small code as example but commented:
```suggestion
    // Do whatever you need after a successful upgrade action
    // const moduleElement = $(event.detail[0]);
    // const techName = moduleElement.data('tech-name');
```

Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: Cleaning it means clean the event listener in this controller. Useless 😄 

==> Done

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (kpodemski) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: ok, I didn't notice that you had to copy the code from one file to another

do you think you could export this code and import in both places? 🤔

File: admin-dev/themes/new-theme/js/pages/module/controller.js
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/pages/module/controller.js]: You're absolutely right. It can be useful

**Review by kpodemski — DISMISSED** [review]

File: admin-dev/themes/new-theme/js/components/bo-event-emitter.ts
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/components/bo-event-emitter.ts]: I think you misunderstood my comment 😅
The possible solutions were either:
- keep using the initial `BOEvent` but no need to define it twice since it was already done in module-card.ts and accessible in controller.js via window.BOEvent
- use the already existing EventEmitter component everywhere, although we can't be sure it's not a breaking change so risky move

You mixed both solutions, but there's no reason to add a new event component since we already have EventEmitter

My suggestion for safety is to keep using the initial BOEvent for now, unless we're sure that no module rely on it and it's purely internal In which case you can use the existing EventEmitter

**Review by jolelievre — CHANGES_REQUESTED** [review]

File: admin-dev/themes/new-theme/js/components/bo-event-emitter.ts
Comment (sowbiba) [reply admin-dev/themes/new-theme/js/components/bo-event-emitter.ts]: The one called in controller.js : window.BOEvent is the [legacy one](https://github.com/PrestaShop/PrestaShop/blob/develop/admin-dev/themes/default/js/bundle/default.js#L68). It doesnt have the 3rd parameter `datas`
That's why I redefined it previously in controller.js

To have it in common in the controller and in module-card, I choosed to define a single component and import it

I will try to call BOEvent and not window.BOEvent to see if it's accessible like this


EDIT : I confirm, window.BOEvent or just BOEvent, it's the same, they rely on the legacy component. If I want to keep it in module-card.ts, I have to remove the ```export default class ModuleCard```, export both ModuleCard and BOEvent from that file and import BOEvent in controller.js
Which is the same as create a dedicated file for the component BOEvent 

EDIT 2 : I prefer to use the eventEmitter in window.prestashop.instance.eventEmitter instead

**Review by jolelievre — APPROVED** [review]
Thanks @sowbiba 

Perfect!

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
@PrestaShop/qa-functional @jolelievre I think this one should be tested by a developer, right?

**Comment by Robin-Fischer-PS** [comment]
Unless if someone find an How to test not too technical, I agree with @kpodemski !

**Comment by WahbiPS** [comment]
What do we need to test this PR @sowbiba ? one of your modules maybe ?

**Comment by sowbiba** [comment]
Example modules updated

1 - Install [prestashopexamplemodule_after_upgrade_1.0.0.zip](https://github.com/PrestaShop/PrestaShop/files/11154372/prestashopexamplemodule_after_upgrade_1.0.0.zip). You'll have in the configration table an entry PRESTASHOPEXAMPLEMODULE_DEBUG with null value


2 - Import the upgrade version [prestashopexamplemodule_after_upgrade_1.0.1.zip](https://github.com/PrestaShop/PrestaShop/files/11154379/prestashopexamplemodule_after_upgrade_1.0.1.zip). You'll have an ajax call after the module import and the value of PRESTASHOPEXAMPLEMODULE_DEBUG will be : Upgraded at Y/m/d H:i:s

You can see the ajax call in the JS and network consoles

**Comment by matks** [comment]
> Example modules updated
> 
> 1 - Install [prestashopexamplemodule_after_upgrade_1.0.0.zip](https://github.com/PrestaShop/PrestaShop/files/11154372/prestashopexamplemodule_after_upgrade_1.0.0.zip). You'll have in the configration table an entry PRESTASHOPEXAMPLEMODULE_DEBUG with null value
> 
> 2 - Import the upgrade version [prestashopexamplemodule_after_upgrade_1.0.1.zip](https://github.com/PrestaShop/PrestaShop/files/11154379/prestashopexamplemodule_after_upgrade_1.0.1.zip). You'll have an ajax call after the module import and the value of PRESTASHOPEXAMPLEMODULE_DEBUG will be : Upgraded at Y/m/d H:i:s
> 
> You can see the ajax call in the JS and network consoles

I confirm this behavior 👍 

Capture d’écran 2023-04-06 à 12 06 42

**Comment by matks** [comment]
For me the PR is OK 👍 I have asked some more regular tests from QA team because this is quite a critical part of the software 😄

**Comment by sowbiba** [comment]
Perfect. Thanks @matks 
Waiting for the QA feedback 😉

**Review by florine2623 — APPROVED** [review]
Hello @sowbiba ,

Thanks for the PR! 

I followed [your steps](https://github.com/PrestaShop/PrestaShop/pull/31285#issuecomment-1496805123), I'm not having the same result as @matks .

After the first import of module 1.0.0 :
Screenshot 2023-04-07 at 14 37 07

After the import of the upgraded module 1.0.1 :
Screenshot 2023-04-07 at 14 37 37

On the first try, I don't see the `upgrade` call as in : https://github.com/PrestaShop/PrestaShop/pull/31285#issuecomment-1498809455

Also, when I try to reimport the modules to retest a second time, the BO is blocked :

https://user-images.githubusercontent.com/16019289/230611323-021a8942-8f2b-4981-8bc9-890f30237ccd.mov

But after deleting, reimporting the test module, it works :

https://user-images.githubusercontent.com/16019289/230611523-6c4837ad-d029-458f-a56a-87c2b274b0bc.mov

---

Also tested the import of other modules, tested the actions buttons of module page, checked Alerts and Updates tab.

It is QA ✅

**Comment by nicosomb** [comment]
Thank you @sowbiba !

**Pull request merged**
